### PR TITLE
CNF-6517: [Part 2] Hypershift PAO adoption

### PR DESCRIPTION
**Pull request by Tal-or**
Since the controller implementation diverged on Hypershift platform, we need to add an abstraction layer
in order to handle a different type of implementation.

In this PR we introduce 2 interfaces:
1. `components.Handler` - which provides an API for handling the different components the controller owns.
2. `status.Writer` which provides an API for reporting the different statuses by the controller.

This PR doesn't add any new functionality but implementing the old reconciliation loop using the new interfaces introduced above.

### CODE REVIEW

**Comment by Tal-or** [comment]
/test e2e-gcp-pao
```
Falling back to `oc adm inspect clusteroperators.v1.config.openshift.io` to collect basic cluster information.
  Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
  error running backup collection: inspection completed with the errors occurred while gathering data:
      [skipping gathering secrets/support due to error: secrets "support" not found, skipping gathering endpoints/host-etcd-2 due to error: endpoints "host-etcd-2" not found, skipping gathering sharedconfigmaps.sharedresource.openshift.io due to error: the server doesn't have a resource type "sharedconfigmaps", skipping gathering sharedsecrets.sharedresource.openshift.io due to error: the server doesn't have a resource type "sharedsecrets"]
  error: gather did not start for pod must-gather-d7xjp: unable to pull image: ImagePullBackOff: Back-off pulling image "registry.build03.ci.openshift.org/ci-op-z8lw46nb/stable@sha256:addc9b013a4cbbe1067d652d032048e7b5f0c867174671d51cbf55f765ff35fc"
  [FAILED] in [BeforeSuite] - /go/src/github.com/openshift/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/6_mustgather_testing/test_suite_mustgather_test.go:42 @ 04/18/24 10:12:58.063
[BeforeSuite] [FAILED] [187.436 seconds]
[BeforeSuite] 
/go/src/github.com/openshift/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/6_mustgather_testing/test_suite_mustgather_test.go:20
  [FAILED] Unexpected error:
      <*exec.ExitError | 0xc00014c060>: 
      exit status 1
      {
          ProcessState: {
              pid: 13528,
              status: 256,
              rusage: {
                  Utime: {Sec: 39, Usec: 823355},
                  Stime: {Sec: 9, Usec: 256492},
                  Maxrss: 193088,
                  Ixrss: 0,
                  Idrss: 0,
                  Isrss: 0,
                  Minflt: 459497,
                  Majflt: 2,
                  Nswap: 0,
                  Inblock: 3808,
                  Oublock: 214320,
                  Msgsnd: 0,
                  Msgrcv: 0,
                  Nsignals: 0,
                  Nvcsw: 209038,
                  Nivcsw: 3664,
              },
          },
          Stderr: nil,
      }
  occurred
  In [BeforeSuite] at: /go/src/github.com/openshift/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/6_mustgather_testing/test_suite_mustgather_test.go:42 @ 04/18/24 10:12:58.063
```

**Comment by Tal-or** [comment]
Depends on #1032

File: pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	performancev2 "github.com/openshift/cluster-node-tuning-operator/pkg/apis/performanceprofile/v2"
+	"github.com/openshift/cluster-node-tuning-operator/pkg/performanceprofile/controller/performanceprofile/components"
+	"github.com/openshift/cluster-node-tuning-operator/pkg/performanceprofile/controller/performanceprofile/components/machineconfig"
+	"github.com/openshift/cluster-node-tuning-operator/pkg/performanceprofile/controller/performanceprofile/components/manifestset"
+	profileutil "github.com/openshift/cluster-node-tuning-operator/pkg/performanceprofile/controller/performanceprofile/components/profile"
+	"github.com/openshift/cluster-node-tuning-operator/pkg/performanceprofile/controller/performanceprofile/resources"
+	"github.com/openshift/cluster-node-tuning-operator/pkg/performanceprofile/controller/performanceprofile/status"
+)
+
+var _ components.Handler = &handler{}
```
Comment (rbaturov) [inline pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go:29]: Just wondering, why is this needed?

File: pkg/performanceprofile/controller/performanceprofile/status/status.go [hunk 2]
```diff
@@ -32,6 +32,13 @@ const (
 	ConditionFailedGettingTunedProfileStatus = "GettingTunedStatusFailed"
 )
 
+type Writer interface {
+	// Update updates the status reported by the controller
+	Update(ctx context.Context, object client.Object, conditions []conditionsv1.Condition) error
+	// UpdateOwnedConditions updates the conditions of the components owned by the controller
+	UpdateOwnedConditions(ctx context.Context, object client.Object, profileMCP *mcov1.MachineConfigPool) error
```
Comment (rbaturov) [inline pkg/performanceprofile/controller/performanceprofile/status/status.go:38]: In the implementation for `UpdateOwnedConditions` in hypershift, are we counting on calling this function with an empty  `*mcov1.MachineConfigPool` (we said no MCP on hypershift IIRC)?

File: pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go (hunk 1 above)
Comment (ffromani) [reply pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go:29]: it's because https://go.dev/doc/faq#guarantee_satisfies_interface

File: pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+type handler struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+func NewHandler(cli client.Client, scheme *runtime.Scheme) components.Handler {
+	return &handler{Client: cli, scheme: scheme}
+}
+
+func (h *handler) Apply(ctx context.Context, obj client.Object, recorder record.EventRecorder, opts *components.Options) (*reconcile.Result, error) {
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go:40]: it seems to me that we either return `nil, err` or `reconcile.Result{}, nil`. If so I'd remove the `reconcile.Result` return value which unnecessarily ties the API to the reconcile loop. We can perhaps just return `error` and make logic in the call site. It seems to me just returning `error` gives the same amount of information.

File: pkg/performanceprofile/controller/performanceprofile/status/status.go (hunk 2 above)
Comment (ffromani) [reply pkg/performanceprofile/controller/performanceprofile/status/status.go:38]: this is a good point indeed

File: pkg/performanceprofile/controller/performanceprofile/status/writer.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+func (w *writer) Update(ctx context.Context, object client.Object, conditions []conditionsv1.Condition) error {
+	profile, ok := object.(*performancev2.PerformanceProfile)
+	if !ok {
+		return fmt.Errorf("wrong type conversion; want=PerformanceProfile got=%T", object)
+	}
+	return w.update(ctx, profile, conditions)
+}
+
+func (w *writer) UpdateOwnedConditions(ctx context.Context, object client.Object, profileMCP *mcov1.MachineConfigPool) error {
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/status/writer.go:33]: how hard it is (perhaps in a later PR) to fetch the correct MCP inside this function instead of requiring as argument? Reason I'm asking is to look for consistent API between hypershift and non-hypershift. Per previous comments, looks like this argument should always be `nil` in hypershift, which suggest the API is not generic enough.

File: pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	profile, ok := obj.(*performancev2.PerformanceProfile)
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go:41]: I guess in the hypershift implementation we will need a `ConfigMap` object and we will bail out if it is not a `ConfigMap` object, right?

File: pkg/performanceprofile/controller/performanceprofile_controller.go [hunk 6]
```diff
@@ -683,6 +455,34 @@ func hasFinalizer(profile *performancev2.PerformanceProfile, finalizer string) b
 	return false
 }
 
+func (r *PerformanceProfileReconciler) getAndValidateMCP(ctx context.Context, instance client.Object) (*mcov1.MachineConfigPool, *reconcile.Result, error) {
```
Comment (ffromani) [inline pkg/performanceprofile/controller/performanceprofile_controller.go:457]: can we push this function (or an amended form of) inside `err = r.StatusWriter.UpdateOwnedConditions(ctx, instance, profileMCP)`, so we can remove the potentially problematic `profileMCP` argument?

**Review by ffromani — COMMENTED** [review]
When adding interfaces or generalizing APIs, a painful lesson over time is that we almost always benefit from seeing at least two implementation of the same interface, to be reasonnably confident the abstraction is correct and generic enough.

Here we see just one, but somehow the proposed interfaces click right. So I'm positive about moving on, with relatively minor suggestions inside

But still, let me ask: there is any chance to see a sketch of how the hypershift implementation of the interfaces introduced here could look like?

File: pkg/performanceprofile/controller/performanceprofile/status/status.go (hunk 2 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/status/status.go:38]: Another option is to get the MCP from within the function

**Comment by Tal-or** [comment]
@ffromani 
> But still, let me ask: there is any chance to see a sketch of how the hypershift implementation of the interfaces introduced here could look like?

I'm sorry if it's going to be too much of hand waving but i'll try to point out the general flow:
1. The object we expect to have is always a `ConfigMap` and not a `PerformanceProfile` as we see here.
2. No MCPs so wherever MCP requested we should pass `nil`
3. For checking the existence of object the operator looks for  for a `ConfigMap`s inside its namespace which encapsulated the owned objects i.e `MachineConfig`, `KubeletConfig`. `RuntimeClass` object is a different case and the operator checks for its existence on the hostedcluster.
4. The apply flows goes similar as described above. the operator creates the objects inside its namespace and encapsulate them in a `ConfigMap`. 

Another sketch can be seen in the old PR here: 
https://github.com/openshift/cluster-node-tuning-operator/pull/484

File: pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go (hunk 5 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go:41]: Yes. Exactly

File: pkg/performanceprofile/controller/performanceprofile_controller.go (hunk 6 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile_controller.go:457]: Sure we do, sounds like a good idea

File: pkg/performanceprofile/controller/performanceprofile/status/status.go (hunk 2 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/status/status.go]: Done

File: pkg/performanceprofile/controller/performanceprofile/status/writer.go (hunk 4 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/status/writer.go]: Done

File: pkg/performanceprofile/controller/performanceprofile_controller.go (hunk 6 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile_controller.go:457]: I removed the `ProfileMCP` from the `UpdateOwnedConditions` as requested. Good idea, thank you

File: pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go (hunk 3 above)
Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go:40]: OK you were right, it doesn't matter, I thought `ctrl.Result{}` and  `reconcile.Result{}` are different, but they are the same

Comment (Tal-or) [reply pkg/performanceprofile/controller/performanceprofile/components/handler/handler.go]: Done

**Review by ffromani — COMMENTED** [review]
/approve
/lgtm

thanks for the updates

**Comment by ffromani** [comment]
> @ffromani
> 
> > But still, let me ask: there is any chance to see a sketch of how the hypershift implementation of the interfaces introduced here could look like?
> 
> I'm sorry if it's going to be too much of hand waving but i'll try to point out the general flow:
> 
>     1. The object we expect to have is always a `ConfigMap` and not a `PerformanceProfile` as we see here.
> 
>     2. No MCPs so wherever MCP requested we should pass `nil`
> 
>     3. For checking the existence of object the operator looks for  for a `ConfigMap`s inside its namespace which encapsulated the owned objects i.e `MachineConfig`, `KubeletConfig`. `RuntimeClass` object is a different case and the operator checks for its existence on the hostedcluster.
> 
>     4. The apply flows goes similar as described above. the operator creates the objects inside its namespace and encapsulate them in a `ConfigMap`.
> 
> 
> Another sketch can be seen in the old PR here: #484

good enough for now, let's move on

**Comment by Tal-or** [comment]
/retest

infra issue

**Comment by ffromani** [comment]
/retest-required

it seems the hypershift lane is flaky (again)
